Move from python 2 to python 3

razzeee opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->

This is mostly not my work, I'm just creating the PR.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->
Has been done by our gsoc two years ago, as python 2 end of life is approaching.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->
Has been tested multiple times by us and members of the community.

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [x] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)


# List of things that still need to be done:

- [ ] Check/Fix UWP Platform support
- [ ] This is using the migration repos for now, we probably want to switch that to matrix

Please feel free to add more todos.

arpitn30 committed:
python3 implementation

* Updated all the files from old function names to new function names as
  given in pydocs - http://py3c.readthedocs.io/en/latest/reference.html
* changed char to wchar_t
* File pointer (FILE *) updated
* Updated ob_type to be called by multi layered PyObject
* Removed tp_compare since it's depreciated
* Updated definitions for InitModule
* Updated the Python3 Android Script
* Updated the GIL acquisition and release statements
* Updated module initialization from init<module> to PyInit_<module> and added the new modules to built-in library through PyImport_AppendInitTab
* Charset Conversion From char To wchar_t
* Updated Python WSGI Invoker

arpitn30 committed:
Converted Virtual to Pure Virtual Functions

arpitn30 committed:
Updated Windows Dependencies

arpitn30 committed:
[depends] Native Files Updated

arpitn30 committed:
[depends] Updated Python3 Makefile and Patches

arpitn30 committed:
[depends] Updated the PythonPath in target directories

arpitn30 committed:
Added IOS and OSX Fix

arpitn30 committed:
Added Android Fix

arpitn30 committed:
Added Python Version Varaiable

Rechi committed:
[debian][cpack] switch to python3 depends

Rechi committed:
[cleanup] remove unused functions

Rechi committed:
[cleanup] remove unused parameters from executeScript function

peak3d committed:
Revert "Reusepython"

Rechi committed:
[TEMP] add python BASE_URL

Rechi committed:
[TEMP] add windows precompiled python lib

lrusak commented:
Thanks for picking this up. So the only thing holding us back is UWP?

Razzeee commented:
I think so, at least that's the only thing that  I'm aware of. And it makes sense, as we did not have that platform back then.

afedchin commented:
before I will able provide compiled py3-uwp packages I have to synchronize dependencies between desktop and uwp. It means that at least 3 dependencies should be updated (see https://github.com/xbmc/xbmc/pull/14777/files#diff-ee1acefffbd00c4e123c845b4ef158f0). also I don't know were windows dependencies built with updated OpenSSL or not. @Rechi?

afedchin commented:
uploaded `python3-3.7.2-win10-x64-v141.7z` and `python3-3.7.2-win10-Win32-v141.7z` to mirrors

faush01 commented:
It would be awesome to see this, not that I am blocked but it would be a big step forward to see python 3 in the Kodi 19 nightly builds.

lrusak commented:
Looks like we are using the depreciated method `PyEval_AcquireLock`. It should probably be updated at some point.

https://docs.python.org/3/c-api/init.html#c.PyEval_AcquireLock

Razzeee commented:
@lrusak I addressed all but one, that one isn't obvious for me.

@afedchin I added your zip files, should that be enough?

Some builds are not working, haven't looked at the new logs. If anybody can help it would be appreciated.

https://jenkins.kodi.tv/job/BuildMulti-PR/12399/

afedchin commented:
> @afedchin I added your zip files, should that be enough?

actually no. packages are broken. I suggest to wait until https://github.com/Rechi/xbmc/tree/windows/depends goes in

lrusak commented:
@Razzeee Thanks for doing that. It looks like the builds are still failing because of the max warnings.

```
01:21:51 [WARNINGS] Plug-in Result: Failed - <a href="warningsResult">7 warnings</a> of <a href="warningsResult/NORMAL">priority Normal</a> exceed the threshold of 6 by 1
01:21:52 Build step '[Deprecated] Scan for compiler warnings' changed build result to FAILURE
```

The last new warning is
```
03:41:33 /home/jenkins/workspace/LINUX-64/xbmc/interfaces/python/XBPython.cpp: In member function ‘virtual bool XBPython::OnScriptInitialized(ILanguageInvoker*)’:
03:41:33 /home/jenkins/workspace/LINUX-64/xbmc/interfaces/python/XBPython.cpp:599:26: warning: ‘void PyEval_AcquireLock()’ is deprecated [-Wdeprecated-declarations]
03:41:33        PyEval_AcquireLock();
03:41:33                           ^
03:41:33 In file included from /home/jenkins/workspace/LINUX-64/tools/depends/xbmc-depends/x86_64-linux-gnu-debug/include/python3.7/Python.h:131:0,
03:41:33                  from /home/jenkins/workspace/LINUX-64/xbmc/interfaces/python/XBPython.cpp:10:
03:41:33 /home/jenkins/workspace/LINUX-64/tools/depends/xbmc-depends/x86_64-linux-gnu-debug/include/python3.7/ceval.h:198:18: note: declared here
03:41:33  PyAPI_FUNC(void) PyEval_AcquireLock(void) Py_DEPRECATED(3.2);
03:41:33                   ^~~~~~~~~~~~~~~~~~```

Hopefully that can be fixed

Razzeee commented:
@lrusak had another look, last new warning should be fixed, but needs testing

Razzeee commented:
Way better, only `UWP`, `TVOS` and `Win 32` builds are failing. I start a build that will get uploaded to the mirrors for us to test the changes.

I should probably drop the temporary migration repo changes? @Rechi

fuzzard commented:
tvos build just needs the following into config.site.in

ac_cv_func_sigaltstack=no

Can throw it in the if test "@platform_os@" = "ios"; then if, or check for appletvtvos and python package. 
For a test build ive just added it after the following. 

https://github.com/xbmc/xbmc/blob/d14a5bf1066d3412f488b2ce9e286fc9b30d2c8f/tools/depends/target/config.site.in#L157

Razzeee commented:
Rebased. Dropped the commit that made it use the `migration` repo, will default to `matrix` now.

Razzeee commented on tools/depends/native/python3/Makefile:
```diff
@@ -7,12 +7,16 @@ LIBNAME=Python
+BASE_URL=https://www.python.org/ftp/python/$(VERSION)
 
-CONFIGURE=./configure --prefix=$(NATIVEPREFIX) --disable-shared --disable-framework --without-ensurepip --without-pymalloc --with-system-ffi
+CONFIGURE=./configure --prefix=$(NATIVEPREFIX) --disable-shared --disable-framework --without-pymalloc
```
This conflicted, can you review @Rechi

Rechi replied:
Why remove ` --with-system-ffi`, which was added in master branch?

Razzeee replied:
Should be resolved with the latest version

Razzeee commented on tools/depends/native/Makefile:
```diff
@@ -22,8 +22,8 @@ ifeq ($(OS),osx)
-  ifeq ($(TARGET_PLATFORM),$(filter $(TARGET_PLATFORM),raspberry-pi gbm))
-    NATIVE += libffi meson ninja python3 setuptools
+  ifeq ($(TARGET_PLATFORM),$(filter $(TARGET_PLATFORM),raspberry-pi aml gbm))
+    NATIVE += meson ninja
```
This conflicted please let me know if this is still how it should be @Rechi

Rechi replied:
Why add `aml` (aml windowing), which was dropped in master branch?

Rechi replied:
Why remove `setuptools`, which was added in master branch?

Razzeee replied:
Should be resolved with the latest version

Razzeee commented:
So win32, UWP and TVOS are still failing.

@fuzzard would you be able to link me to the needed code or provide a diff?

Rechi commented on tools/depends/native/Makefile:
```diff
@@ -47,8 +47,8 @@ libpng: zlib
+distribute: python3
+distutilscross: python3 distribute
 tar: xz automake
 python27: zlib
```
removal of this line got lost on your previous rebase

fuzzard commented:
the following is probably the easiest simple fix targeting tvos only.

```
--- a/tools/depends/target/config.site.in
+++ b/tools/depends/target/config.site.in
@@ -171,6 +171,10 @@ if test "@platform_os@" = "ios"; then
   fi
 fi

+#Python3 for appletvos build fix 
+if test "@target_platform@" = "appletvos"
+  ac_cv_func_sigaltstack=no
+fi
+
 #Fix for older, broken android toolchains
 if test "${PACKAGE_NAME}" = "FreeType" -a "@platform_os@" = "android"; then
   CFLAGS="$CFLAGS -std=gnu99"
```

fuzzard commented:
My bad on that one @Razzeee , was missing a ; then at the end of the if statement. Have updated the diff above, that should fix all the build issues failing because of the config.site failing to load

Razzeee commented:
Thanks for letting me know. Hopefully fixed that now, let's wait for jenkins.

Razzeee commented:
@fuzzard osx seems to be broken now, might be due to that change?

Rechi commented:
@Razzeee no, look into the build log to see what actually happened.

Razzeee commented:
jenkins build this please

Razzeee commented:
jenkins build this please

Rechi commented on xbmc/interfaces/python/PythonInvoker.h:
```diff
@@ -9,16 +9,10 @@
 
-#include <map>
-#include <string>
-#include <vector>
```
removing this (three) includes is a rebase error

Razzeee replied:
Should be resolved with the latest version

Rechi commented on tools/depends/target/python3/Makefile:
```diff
@@ -34,31 +35,30 @@ $(PLATFORM): $(TARBALLS_LOCATION)/$(ARCHIVE) $(DEPS)
-	cd $(PLATFORM); sed -ie 's|_locale _localemodule.c   -lintl|_locale _localemodule.c   -lintl $(LINK_ICONV) |' Modules/Setup.dist
+	cd $(PLATFORM); sed -ie 's|_locale _localemodule.c   -lintl|_locale _localemodule.c   -lintl $(LINK_ICONV)|' Modules/Setup.dist
+endif
+ifeq ($(OS),ios)
```
`ifeq ($(OS),darwin_embedded)`

Rechi commented on xbmc/interfaces/generic/ILanguageInvocationHandler.h:
```diff
@@ -24,7 +24,7 @@ class ILanguageInvocationHandler
   virtual void OnScriptStarted(ILanguageInvoker *invoker) { }
   virtual void OnScriptAbortRequested(ILanguageInvoker *invoker) { }
-  virtual void OnExecutionEnded(ILanguageInvoker *invoker) { }
+  virtual void OnScriptEnded(ILanguageInvoker *invoker) { }
```
left align `*`

Rechi commented on xbmc/interfaces/generic/LanguageInvokerThread.cpp:
```diff
@@ -10,33 +10,26 @@
 #include "ScriptInvocationManager.h"
 
-CLanguageInvokerThread::CLanguageInvokerThread(LanguageInvokerPtr invoker, CScriptInvocationManager *invocationManager, bool reuseable)
+CLanguageInvokerThread::CLanguageInvokerThread(LanguageInvokerPtr invoker, CScriptInvocationManager *invocationManager)
```
left align `*`

Rechi commented on xbmc/interfaces/generic/LanguageInvokerThread.h:
```diff
@@ -19,15 +19,10 @@ class CScriptInvocationManager;
 {
 public:
-  CLanguageInvokerThread(LanguageInvokerPtr invoker, CScriptInvocationManager *invocationManager, bool reuseable);
+  CLanguageInvokerThread(LanguageInvokerPtr invoker, CScriptInvocationManager *invocationManager);
```
left align `*`

Rechi commented on xbmc/interfaces/generic/ScriptInvocationManager.cpp:
```diff
@@ -180,51 +175,20 @@ bool CScriptInvocationManager::HasLanguageInvoker(const std::string &script) con
 }
 
-int CScriptInvocationManager::GetReusablePluginHandle(const std::string &script)
+LanguageInvokerPtr CScriptInvocationManager::GetLanguageInvoker(const std::string &script) const
```
left align `&`

Rechi commented on xbmc/interfaces/generic/ScriptInvocationManager.cpp:
```diff
@@ -180,51 +175,20 @@ bool CScriptInvocationManager::HasLanguageInvoker(const std::string &script) con
-  const std::vector<std::string> &arguments /* = std::vector<std::string>() */,
-  bool reuseable /* = false */,
-  int pluginHandle /* = -1 */)
+int CScriptInvocationManager::ExecuteAsync(const std::string &script, const ADDON::AddonPtr &addon /* = ADDON::AddonPtr() */, const std::vector<std::string> &arguments /* = std::vector<std::string>() */)
```
left align `&`

Rechi commented on xbmc/interfaces/generic/ScriptInvocationManager.cpp:
```diff
@@ -236,15 +200,10 @@ int CScriptInvocationManager::ExecuteAsync(const std::string &script,
-  const std::vector<std::string> &arguments /* = std::vector<std::string>() */,
-  bool reuseable /* = false */,
-  int pluginHandle /* = -1 */)
+int CScriptInvocationManager::ExecuteAsync(const std::string &script, LanguageInvokerPtr languageInvoker, const ADDON::AddonPtr &addon /* = ADDON::AddonPtr() */, const std::vector<std::string> &arguments /* = std::vector<std::string>() */)
```
left align `&`

Rechi commented on xbmc/interfaces/generic/ScriptInvocationManager.cpp:
```diff
@@ -255,47 +214,26 @@ int CScriptInvocationManager::ExecuteAsync(const std::string &script,
-  const std::vector<std::string> &arguments /* = std::vector<std::string>() */,
-  uint32_t timeoutMs /* = 0 */,
-  bool waitShutdown /* = false */)
+int CScriptInvocationManager::ExecuteSync(const std::string &script, const ADDON::AddonPtr &addon /* = ADDON::AddonPtr() */, const std::vector<std::string> &arguments /* = std::vector<std::string>() */, uint32_t timeoutMs /* = 0 */, bool waitShutdown /* = false */)
```
left align `&`

Rechi commented on xbmc/interfaces/generic/ScriptInvocationManager.cpp:
```diff
@@ -310,12 +248,7 @@ int CScriptInvocationManager::ExecuteSync(const std::string &script,
-  const std::vector<std::string> &arguments /* = std::vector<std::string>() */,
-  uint32_t timeoutMs /* = 0 */,
-  bool waitShutdown /* = false */)
+int CScriptInvocationManager::ExecuteSync(const std::string &script, LanguageInvokerPtr languageInvoker, const ADDON::AddonPtr &addon /* = ADDON::AddonPtr() */, const std::vector<std::string> &arguments /* = std::vector<std::string>() */, uint32_t timeoutMs /* = 0 */, bool waitShutdown /* = false */)
```
left align `&`

Rechi commented on xbmc/interfaces/generic/ScriptInvocationManager.h:
```diff
@@ -32,12 +32,7 @@ class CScriptInvocationManager
-  * \brief Returns addon_handle if last reusable invoker is ready to use.
-  */
-  int GetReusablePluginHandle(const std::string &script);
+  LanguageInvokerPtr GetLanguageInvoker(const std::string &script) const;
```
left align `&`

Rechi commented on xbmc/interfaces/generic/ScriptInvocationManager.h:
```diff
@@ -47,11 +42,7 @@ class CScriptInvocationManager
-    const std::vector<std::string> &arguments = std::vector<std::string>(),
-    bool reuseable = false,
-    int pluginHandle = -1);
+  int ExecuteAsync(const std::string &script, const ADDON::AddonPtr &addon = ADDON::AddonPtr(), const std::vector<std::string> &arguments = std::vector<std::string>());
```
left align `&`

Rechi commented on xbmc/interfaces/generic/ScriptInvocationManager.h:
```diff
@@ -61,12 +52,7 @@ class CScriptInvocationManager
-    const std::vector<std::string> &arguments = std::vector<std::string>(),
-    bool reuseable = false,
-    int pluginHandle = -1);
+  int ExecuteAsync(const std::string &script, LanguageInvokerPtr languageInvoker, const ADDON::AddonPtr &addon = ADDON::AddonPtr(), const std::vector<std::string> &arguments = std::vector<std::string>());
```
left align `&`

Rechi commented on xbmc/interfaces/generic/ScriptInvocationManager.h:
```diff
@@ -84,11 +70,7 @@ class CScriptInvocationManager
-    const std::vector<std::string> &arguments = std::vector<std::string>(),
-    uint32_t timeoutMs = 0,
-    bool waitShutdown = false);
+  int ExecuteSync(const std::string &script, const ADDON::AddonPtr &addon = ADDON::AddonPtr(), const std::vector<std::string> &arguments = std::vector<std::string>(), uint32_t timeoutMs = 0, bool waitShutdown = false);
```
left align `&`

Rechi commented on xbmc/interfaces/generic/ScriptInvocationManager.h:
```diff
@@ -106,12 +88,7 @@ class CScriptInvocationManager
-    const std::vector<std::string> &arguments = std::vector<std::string>(),
-    uint32_t timeoutMs = 0,
-    bool waitShutdown = false);
+  int ExecuteSync(const std::string &script, LanguageInvokerPtr languageInvoker, const ADDON::AddonPtr &addon = ADDON::AddonPtr(), const std::vector<std::string> &arguments = std::vector<std::string>(), uint32_t timeoutMs = 0, bool waitShutdown = false);
```
left align `&`

Rechi commented on xbmc/interfaces/python/PythonInvoker.cpp:
```diff
@@ -155,121 +156,115 @@ bool CPythonInvoker::execute(const std::string &script, const std::vector<std::s
-    PyObject *pathObj(PyDict_GetItemString(sysModDict, "path")); // borrowed ref, no need to delete
+  // we want to use sys.path so it includes site-packages
+  // if this fails, default to using Py_GetPath
+  PyObject *sysMod(PyImport_ImportModule("sys")); // must call Py_DECREF when finished
```
left align `*`

Rechi commented on xbmc/interfaces/python/PythonInvoker.cpp:
```diff
@@ -155,121 +156,115 @@ bool CPythonInvoker::execute(const std::string &script, const std::vector<std::s
+  // we want to use sys.path so it includes site-packages
+  // if this fails, default to using Py_GetPath
+  PyObject *sysMod(PyImport_ImportModule("sys")); // must call Py_DECREF when finished
+  PyObject *sysModDict(PyModule_GetDict(sysMod)); // borrowed ref, no need to delete
```
left align `*`

Rechi commented on xbmc/interfaces/python/PythonInvoker.cpp:
```diff
@@ -155,121 +156,115 @@ bool CPythonInvoker::execute(const std::string &script, const std::vector<std::s
+  // if this fails, default to using Py_GetPath
+  PyObject *sysMod(PyImport_ImportModule("sys")); // must call Py_DECREF when finished
+  PyObject *sysModDict(PyModule_GetDict(sysMod)); // borrowed ref, no need to delete
+  PyObject *pathObj(PyDict_GetItemString(sysModDict, "path")); // borrowed ref, no need to delete
```
left align `*`

Rechi commented on xbmc/interfaces/python/PythonInvoker.cpp:
```diff
@@ -155,121 +156,115 @@ bool CPythonInvoker::execute(const std::string &script, const std::vector<std::s
-        PyObject *e = PyList_GetItem(pathObj, i); // borrowed ref, no need to delete
-        if (e != NULL && PyString_Check(e))
-          addNativePath(PyString_AsString(e)); // returns internal data, don't delete or modify
+      PyObject *e = PyList_GetItem(pathObj, i); // borrowed ref, no need to delete
```
left align `*`

Rechi commented on xbmc/interfaces/python/PythonInvoker.cpp:
```diff
@@ -155,121 +156,115 @@ bool CPythonInvoker::execute(const std::string &script, const std::vector<std::s
+      // uwp python operates unicodes
+      else if (e != NULL && PyUnicode_Check(e))
+      {
+        PyObject *utf8 = PyUnicode_AsUTF8String(e);
```
left align `*`

Rechi commented on xbmc/interfaces/python/PythonInvoker.cpp:
```diff
@@ -357,50 +352,83 @@ bool CPythonInvoker::execute(const std::string &script, const std::vector<std::s
-    PyObject *m = PyImport_AddModule("xbmc");
-    if (m == NULL || PyObject_SetAttrString(m, "abortRequested", PyBool_FromLong(1)))
-      CLog::Log(LOGERROR, "CPythonInvoker(%d, %s): failed to set abortRequested", GetId(), m_sourceFile.c_str());
+  PyObject *m = PyImport_AddModule("xbmc");
```
left align `*`

Rechi commented on xbmc/interfaces/python/PythonInvoker.cpp:
```diff
@@ -357,50 +352,83 @@ bool CPythonInvoker::execute(const std::string &script, const std::vector<std::s
+  // if not obeyed there is still no deadlock because ::stop waits with timeout (smart one!)
+  m_stoppedEvent.Set();
+
+  { CSingleLock lock(m_critical);
```
move curly bracket into a separate line

Rechi commented on xbmc/interfaces/python/PythonInvoker.cpp:
```diff
@@ -420,38 +448,28 @@ bool CPythonInvoker::stop(bool abort)
-      m = PyImport_AddModule("xbmc");
-      if (m == NULL || PyObject_SetAttrString(m, "abortRequested", PyBool_FromLong(1)))
-        CLog::Log(LOGERROR, "CPythonInvoker(%d, %s): failed to set abortRequested", GetId(), m_sourceFile.c_str());
+    PyObject *m;
```
left align `*`

Rechi commented on xbmc/interfaces/python/PythonInvoker.h:
```diff
@@ -63,12 +58,9 @@ class CPythonInvoker : public ILanguageInvoker
 
   std::string m_pythonPath;
-  _ts *m_threadState;
+  void *m_threadState;
```
left align `*`

Rechi commented on xbmc/interfaces/python/XBPython.cpp:
```diff
@@ -680,7 +680,7 @@ void XBPython::OnScriptAbortRequested(ILanguageInvoker *invoker)
 }
 
-void XBPython::OnExecutionEnded(ILanguageInvoker *invoker)
+void XBPython::OnScriptEnded(ILanguageInvoker *invoker)
```
left align `*`

Rechi commented on xbmc/interfaces/python/XBPython.h:
```diff
@@ -91,7 +91,7 @@ class XBPython :
   void OnScriptStarted(ILanguageInvoker *invoker) override;
   void OnScriptAbortRequested(ILanguageInvoker *invoker) override;
-  void OnExecutionEnded(ILanguageInvoker *invoker) override;
+  void OnScriptEnded(ILanguageInvoker *invoker) override;
```
left align `*`

Rechi commented on xbmc/interfaces/python/PythonInvoker.cpp:
```diff
@@ -428,13 +428,13 @@ bool CPythonInvoker::execute(const std::string &script, const std::vector<std::s
 }
 
-void CPythonInvoker::executeScript(void *fp, const std::string &script, void *module, void *moduleDict)
+void CPythonInvoker::executeScript(FILE *fp, const std::string &script, PyObject *moduleDict)
```
left align `*` and `&`

Rechi commented on xbmc/interfaces/python/PythonInvoker.h:
```diff
@@ -31,7 +31,7 @@ class CPythonInvoker : public ILanguageInvoker
   // implementation of ILanguageInvoker
   bool execute(const std::string &script, const std::vector<std::string> &arguments) override;
-  virtual void executeScript(void *fp, const std::string &script, void *module, void *moduleDict);
+  virtual void executeScript(FILE *fp, const std::string &script, PyObject* moduleDict);
```
left align `*` and `&`

Rechi commented on xbmc/network/httprequesthandler/python/HTTPPythonWsgiInvoker.cpp:
```diff
@@ -107,10 +107,10 @@ HTTPPythonRequest* CHTTPPythonWsgiInvoker::GetRequest()
 }
 
-void CHTTPPythonWsgiInvoker::executeScript(void *fp, const std::string &script, void *module, void *moduleDict)
+void CHTTPPythonWsgiInvoker::executeScript(FILE* fp, const std::string &script, PyObject* moduleDict)
```
left align `&`

Rechi commented on xbmc/network/httprequesthandler/python/HTTPPythonWsgiInvoker.h:
```diff
@@ -34,7 +34,7 @@ class CHTTPPythonWsgiInvoker : public CHTTPPythonInvoker
 protected:
   // overrides of CPythonInvoker
-  void executeScript(void *fp, const std::string &script, void *module, void *moduleDict) override;
+  void executeScript(FILE* fp, const std::string &script, PyObject* moduleDict) override;
```
left align `&`

Rechi commented on xbmc/interfaces/python/AddonPythonInvoker.cpp:
```diff
@@ -64,12 +64,12 @@
+  PyObject *PyInit_Module_xbmc(void);
+  PyObject *PyInit_Module_xbmcplugin(void);
+  PyObject *PyInit_Module_xbmcaddon(void);
+  PyObject *PyInit_Module_xbmcvfs(void);
```
do not indent namespaces and left align `*`

Rechi commented on xbmc/interfaces/python/AddonPythonInvoker.cpp:
```diff
@@ -82,17 +82,24 @@ typedef struct
+    PyImport_AppendInittab("xbmc",       PyInit_Module_xbmc);
+    PyImport_AppendInittab("xbmcplugin", PyInit_Module_xbmcplugin);
+    PyImport_AppendInittab("xbmcaddon",  PyInit_Module_xbmcaddon);
+    PyImport_AppendInittab("xbmcvfs",    PyInit_Module_xbmcvfs);
```
wrong indentation

Rechi commented on xbmc/interfaces/python/PythonInvoker.cpp:
```diff
@@ -80,23 +78,23 @@ static const std::string getListOfAddonClassesAsString(XBMCAddon::AddonClass::Re
 }
 
-static std::vector<std::vector<char>> storeArgumentsCCompatible(std::vector<std::string> const & input)
+static std::vector<std::vector<wchar_t>> storeArgumentsCCompatible(std::vector<std::wstring> const & input)
```
left align `&`

Rechi commented on xbmc/interfaces/python/PythonInvoker.cpp:
```diff
@@ -80,23 +78,23 @@ static const std::string getListOfAddonClassesAsString(XBMCAddon::AddonClass::Re
 }
 
-static std::vector<char *> getCPointersToArguments(std::vector<std::vector<char>> & input)
+static std::vector<wchar_t *> getCPointersToArguments(std::vector<std::vector<wchar_t>> & input)
```
left align `*` and `&`

Rechi commented on xbmc/interfaces/python/PythonInvoker.cpp:
```diff
@@ -80,23 +78,23 @@ static const std::string getListOfAddonClassesAsString(XBMCAddon::AddonClass::Re
+static std::vector<wchar_t *> getCPointersToArguments(std::vector<std::vector<wchar_t>> & input)
 {
-  std::vector<char *> output;
+  std::vector<wchar_t *> output;
```
left align `*`

Rechi commented on xbmc/interfaces/python/PythonInvoker.cpp:
```diff
@@ -80,23 +78,23 @@ static const std::string getListOfAddonClassesAsString(XBMCAddon::AddonClass::Re
+  std::vector<wchar_t *> output;
   std::transform(input.begin(), input.end(), std::back_inserter(output),
-                [](std::vector<char> & i) { return &i[0]; });
+                [](std::vector<wchar_t> & i) { return &i[0]; });
```
left align `&`

Rechi commented on xbmc/interfaces/python/PythonInvoker.cpp:
```diff
@@ -139,19 +137,32 @@ bool CPythonInvoker::Execute(const std::string &script, const std::vector<std::s
+  return execute(script, w_arguments);
+}
+
+bool CPythonInvoker::execute(const std::string &script, const std::vector<std::wstring> &arguments)
```
left align `&`

Rechi commented on xbmc/interfaces/python/PythonInvoker.cpp:
```diff
@@ -139,19 +137,32 @@ bool CPythonInvoker::Execute(const std::string &script, const std::vector<std::s
-  std::vector<std::vector<char>> argvStorage = storeArgumentsCCompatible(arguments);
-  std::vector<char *> argv = getCPointersToArguments(argvStorage);
+  std::vector<std::vector<wchar_t>> argvStorage = storeArgumentsCCompatible(arguments);
+  std::vector<wchar_t *> argv = getCPointersToArguments(argvStorage);
```
left align `*`

Rechi commented on xbmc/interfaces/python/PythonInvoker.cpp:
```diff
@@ -278,13 +285,11 @@ bool CPythonInvoker::execute(const std::string &script, const std::vector<std::s
-      //! @bug libpython isn't const correct
-      PyObject* file = PyFile_FromString(const_cast<char*>(nativeFilename.c_str()), const_cast<char*>("r"));
-      FILE *fp = PyFile_AsFile(file);
+      FILE *fp = _Py_fopen(nativeFilename.c_str(), "r");
```
left align `*`

Rechi commented on xbmc/interfaces/python/PythonInvoker.cpp:
```diff
@@ -278,13 +285,11 @@ bool CPythonInvoker::execute(const std::string &script, const std::vector<std::s
       if (fp != NULL)
       {
-        PyObject *f = PyString_FromString(nativeFilename.c_str());
+        PyObject *f = PyUnicode_FromString(nativeFilename.c_str());
```
left align `*`

Rechi commented on xbmc/interfaces/python/PythonInvoker.cpp:
```diff
@@ -437,6 +441,27 @@ void CPythonInvoker::executeScript(FILE *fp, const std::string &script, PyObject
   PyRun_FileExFlags(fp, script.c_str(), m_Py_file_input, moduleDict, moduleDict, 1, NULL);
 }
 
+FILE* CPythonInvoker::PyFile_AsFileWithMode(PyObject *py_file, const char *mode)
```
left align `*`

Rechi commented on xbmc/interfaces/python/PythonInvoker.cpp:
```diff
@@ -437,6 +441,27 @@ void CPythonInvoker::executeScript(FILE *fp, const std::string &script, PyObject
+{
+  PyObject* ret = PyObject_CallMethod(py_file, "flush", "");
+  if (ret == NULL)
+      return NULL;
```
wrong indentation

Rechi commented on xbmc/interfaces/python/PythonInvoker.cpp:
```diff
@@ -560,11 +585,11 @@ void CPythonInvoker::onPythonModuleInitialization(void* moduleDict)
   PyObject *moduleDictionary = (PyObject *)moduleDict;
 
-  PyObject *pyaddonid = PyString_FromString(m_addon->ID().c_str());
+  PyObject *pyaddonid = PyUnicode_FromString(m_addon->ID().c_str());
```
left align `*`

Rechi commented on xbmc/interfaces/python/PythonInvoker.cpp:
```diff
@@ -560,11 +585,11 @@ void CPythonInvoker::onPythonModuleInitialization(void* moduleDict)
 
   ADDON::AddonVersion version = m_addon->GetDependencyVersion("xbmc.python");
-  PyObject *pyxbmcapiversion = PyString_FromString(version.asString().c_str());
+  PyObject *pyxbmcapiversion = PyUnicode_FromString(version.asString().c_str());
```
left align `*`

Rechi commented on xbmc/interfaces/python/swig.cpp:
```diff
@@ -52,18 +52,14 @@ namespace PythonBindings
-        buf = PyString_AsString(utf8_pyString);
-        Py_DECREF(utf8_pyString);
+      buf = PyUnicode_AsUTF8(pObject);
         return;
```
wrong indentation

Rechi commented on xbmc/interfaces/python/swig.cpp:
```diff
@@ -52,18 +52,14 @@ namespace PythonBindings
-    }
-    if (PyString_Check(pObject))
+
+    if (PyBytes_Check(pObject))    // If pobject is of type Bytes
```
wrong indentation

Rechi commented on xbmc/interfaces/python/swig.cpp:
```diff
@@ -167,26 +163,24 @@ namespace PythonBindings
+    if (exc_type != NULL && (pystring = PyObject_Str(exc_type)) != NULL && PyUnicode_Check(pystring))
     {
-      char *str = PyString_AsString(pystring);
+      const char *str = PyUnicode_AsUTF8(pystring);
```
left align `*`

Rechi commented on xbmc/interfaces/python/swig.cpp:
```diff
@@ -199,15 +193,15 @@ namespace PythonBindings
         if (tbList)
         {
-          PyObject *emptyString = PyString_FromString("");
+          PyObject *emptyString = PyUnicode_FromString("");
```
left align `*`

Rechi commented on xbmc/network/httprequesthandler/python/HTTPPythonWsgiInvoker.cpp:
```diff
@@ -6,6 +6,8 @@
  *  See LICENSES/README.md for more information.
  */
 
+#include <Python.h>
```
wrong include order

Rechi commented on xbmc/network/httprequesthandler/python/HTTPPythonWsgiInvoker.cpp:
```diff
@@ -64,9 +66,9 @@
-  void initModule_xbmcwsgi(void);
+  PyObject* PyInit_Module_xbmc(void);
+  PyObject* PyInit_Module_xbmcaddon(void);
+  PyObject* PyInit_Module_xbmcwsgi(void);
```
do not indent namespaces

Rechi commented on xbmc/network/httprequesthandler/python/HTTPPythonWsgiInvoker.cpp:
```diff
@@ -79,15 +81,19 @@ typedef struct
+{
+    PyImport_AppendInittab("xbmc", PyInit_Module_xbmc);
+    PyImport_AppendInittab("xbmcaddon", PyInit_Module_xbmcaddon);
+    PyImport_AppendInittab("xbmcwsgi", PyInit_Module_xbmcwsgi);
```
wrong indentation

enen92 commented:
Would be nice to have this in as soon as possible as a motivation for devs to start porting add-ons to py3.

phunkyfish commented:
Fixes made for iOS/tvOS by @fuzzard.

Runtime tested on iOS/tvOS and OSX.

Razzeee commented:
Tested the build from yesterday on win64 with youtube and twitch. Worked fine.

fuzzard commented:
Testing and runtime verification across all platforms.
Also need some windows dev work for uwp and win32 from jenkins build failures

faush01 commented:
LOL they is a clock :-)
https://pythonclock.org/
Happy to help where I can on this PR, is it just testing needed now?

dagwieers commented:
The Windows Leia Python3 builds are lacking the **script.module.pycryptodome** module (where the normal Leia builds do include this as part of the Kodi distribution). This makes it impossible ATM to get the **plugin.video.netflix** to ship with the Kodi add-ons repository.

candrews commented:
Python 2.7 support ends on January 1 (that's only 3 months and 21 days away) :hourglass_flowing_sand: 

https://pythonclock.org/

faush01 commented:
I spent some time over the last few days to migrating my add-on EmbyCon to Python 3 support using the latest test build from this branch.

https://github.com/faush01/plugin.video.embycon/tree/feature/matrix

I am fortunate in that I have purposely kept to a minimum to number of external add-ons I have used in my add-on thus all the changes needed were internal. This also meant I was able to get the add-on fully functional and able to run Kodi 19 (with Python 3) through its paces.

From what I can tell it looks good. 

I think we need to merge this and get a nightly out there that people can start updating their add-ons for, having add-on owners start looking at this is going to take time and taking this first step with switching the nightly to Python 3 is going to start that process. Yeah some stuff is going to break but the sooner this starts the sooner it is will be done.

Rechi commented on cmake/scripts/darwin_embedded/Install.cmake:
```diff
@@ -51,6 +51,14 @@ add_custom_command(TARGET ${APP_NAME_LC} POST_BUILD
+            "PRODUCT_NAME=${APP_NAME}"
+            "FULL_PRODUCT_NAME=${APP_NAME}.app"
+            "WRAPPER_EXTENSION=app"
+            "SRCROOT=${CMAKE_BINARY_DIR}"
```
looks like a rebase error

jimfcarroll replied:
Why? PYTHON_VERSION now needs to be set so I took the version that contained that. Do you know what it should be?

Rechi replied:
Yes it needs `PYTHON_VERSION`, but that's not in line 54 to 60.
It's in line 61 which is correct and I didn't include in this multi line comment.

jimfcarroll replied:
Okay, but they were a package. Do you want it with just line 61?

Rechi replied:
Have a look at the commit history of this file, this lines where recently removed by intention. You pulled them back while rebasing.

jimfcarroll replied:
okay. When I was resolving the commit conflict I had the choice between all of the lines, or none of the lines. I took the one that included them all because otherwise PYTHON_VERSION wouldn't have been defined, which I saw was needed. That's why they're all there. If you say only PYTHON_VERSION is needed, I'll fix it, but that's why I resolved it that way,

Rechi replied:
Then you are using the wrong tool for rebasing if it only gives you the choice to pick one or the other version and not to interactively fix it.

jimfcarroll replied:
No. You misunderstood me. I could have manually merged them but assuming the state of the code that accompanied the PYTHON_VERSION addition was the correct state was a natural, even if incorrect, assumption.

Rechi replied:
Ah okay.

Please squash the fix commits into the corresponding commits, to not have the wrong changes in the history.

Rechi commented on xbmc/interfaces/python/PythonInvoker.cpp:
```diff
@@ -84,23 +78,23 @@ static const std::string getListOfAddonClassesAsString(XBMCAddon::AddonClass::Re
 
   if (output.empty())
-    output.emplace_back(1u, '\0');
+    output.emplace_back(std::vector<wchar_t>(1u, '\0'));
```
looks like a rebase error

jimfcarroll replied:
Thanks. Will fix

phunkyfish commented:
Jenkins build this please

Commit:
Fix build on various windows by including the correct openssl and Windows UWP by adding -DMS_STORE

phunkyfish commented:
Jenkins build this please

Rechi commented:
There is no reason to tell jenkins to build if it hasn't failed. The reason it didn't start yet is that another PR is currently building.

phunkyfish commented:
Ah, ok. It usually says will build it when I have time.

Rechi commented on xbmc/interfaces/python/swig.cpp:
```diff
@@ -377,5 +377,8 @@ namespace PythonBindings
     return typeInfoLookup[ti];
   }
 
+  int dummy_tp_init(PyObject *self, PyObject *args, PyObject *kwds) {
```
> Left-align `*` and `&` to the base type they modify.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#52-pointer-and-reference-types

> Curly braces always go on a new line.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#31-braces

jimfcarroll replied:
fixed

Rechi commented on xbmc/interfaces/python/swig.h:
```diff
@@ -146,6 +146,8 @@ namespace PythonBindings
   void registerAddonClassTypeInformation(const TypeInfo* classInfo);
   const TypeInfo* getTypeInfoForInstance(XBMCAddon::AddonClass* obj);
 
+  int dummy_tp_init(PyObject *self, PyObject *args, PyObject *kwds);
```
> Left-align `*` and `&` to the base type they modify.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#52-pointer-and-reference-types

jimfcarroll replied:
fixed

Commit:
Fix the ability for subclasses of API classes to call super().__init()__.

ronie commented:
due to this being many months overdue we"re going to take the uwp build failure for granted.

a big thanx to @arpitn30 for this code!

Rechi commented:
This commit still contained temporary commits.

jimfcarroll commented:
Actually, it's not a temporary commit .... At least not anymore

Montellese commented:
Did you merge d8343aacf84026d0a1602d584e59ad09ff71add3 on purpose? Now there are two binaries checked into git.

ronie commented:
i'm not a windows dev, but it'd guess those need to be uploaded to the build-deps on the mirrors..

Montellese commented:
AFAICT that has already been taken care of with 615a55ea but then someone forgot to delete d8343aa. If that's correct then someone should revert that specific commit.

fuzzard commented:
The 2 binaries arent on mirrors, hence why im guessing they were [TEMP] committed for testing.
WindowsStore UWP binaries are on mirrors, but standalone win32 and x64 are not.

Montellese commented:
@fuzzard:You're right. I just checked on the FTP server and `python3-3.7.2-win10-x64-v141.7z` and `python3-3.7.2-win10-Win32-v141.7z` have been uploaded to the `win32` directory but the non-UWP ones not. I wonder however why we use Python 3.7.2 for UWP and 3.7.1 for non-UWP? And for depends we use 3.7.4...

I have upload `python3-3.7.1-Win32-v141.7z` and `python3-3.7.1-x64-v141.7z` to the dependencies mirror.

afedchin commented:
`python3-3.7.2-win10-*` were uploaded by me, but they all aren't working and can be deleted

CastagnaIT commented:
reuselanguageinvoker has been deleted, it will be restored later?
I need to know if i should delete it from the addon

dagwieers commented:
If `reuselanguageinvoker` will disappear, be aware that the overhead of importing the Requests (urllib3) library on every action is slowing down add-ons considerably on Raspberry Pi. It takes about 800ms to import urllib3 on Raspberry Pi.

We migrated our add-ons to using urllib/urllib2 after regressions and performance issues with urllib3.

mediaminister commented:
There is a problem with BeautifulSoup4 in Kodi 19.0-ALPHA1 Git:20191014-05488ed68e on Windows x64: the Python 2 version of BeautifulSoup4 is being used.
I tried to set the dependency of the add-on to the Python 3 version 4.6.3 (http://mirrors.kodi.tv/addons/migration/script.module.beautifulsoup4/script.module.beautifulsoup4-4.6.3.zip)
But I get:
```
2019-10-15 18:37:44.473 T:7256   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: <class 'ModuleNotFoundError'>
                                            Error Contents: No module named 'htmlentitydefs'
                                            Traceback (most recent call last):
                                              File "C:\Users\test\AppData\Roaming\Kodi\addons\plugin.video.vrt.nu\resources\lib\addon_plugin.py", line 8, in <module>
                                                run(sys.argv)
                                              File "C:\Users\test\AppData\Roaming\Kodi\addons\plugin.video.vrt.nu\resources\lib\addon.py", line 293, in run
                                                plugin.run(argv)
                                              File "C:\Users\test\AppData\Roaming\Kodi\addons\script.module.routing\lib\routing.py", line 130, in run
                                                self._dispatch(self.path)
                                              File "C:\Users\test\AppData\Roaming\Kodi\addons\script.module.routing\lib\routing.py", line 141, in _dispatch
                                                view_func(**kwargs)
                                              File "C:\Users\test\AppData\Roaming\Kodi\addons\plugin.video.vrt.nu\resources\lib\addon.py", line 272, in play_url
                                                VRTPlayer(kodi).play(dict(video_url=video_url))
                                              File "C:\Users\test\AppData\Roaming\Kodi\addons\plugin.video.vrt.nu\resources\lib\vrtplayer.py", line 340, in play
                                                stream = _streamservice.get_stream(video)
                                              File "C:\Users\test\AppData\Roaming\Kodi\addons\plugin.video.vrt.nu\resources\lib\streamservice.py", line 193, in get_stream
                                                api_data = self._get_api_data(video)
                                              File "C:\Users\test\AppData\Roaming\Kodi\addons\plugin.video.vrt.nu\resources\lib\streamservice.py", line 105, in _get_api_data
                                                api_data = self._webscrape_api_data(video_url) or ApiData(self._CLIENT, self._VUALTO_API_URL, video_id, '', True)
                                              File "C:\Users\test\AppData\Roaming\Kodi\addons\plugin.video.vrt.nu\resources\lib\streamservice.py", line 110, in _webscrape_api_data
                                                from bs4 import BeautifulSoup, SoupStrainer
                                              File "C:\Users\test\AppData\Roaming\Kodi\addons\script.module.beautifulsoup4\lib\bs4\__init__.py", line 36, in <module>
                                                from .builder import builder_registry, ParserRejectedMarkup
                                              File "C:\Users\test\AppData\Roaming\Kodi\addons\script.module.beautifulsoup4\lib\bs4\builder\__init__.py", line 7, in <module>
                                                from bs4.element import (
                                              File "C:\Users\test\AppData\Roaming\Kodi\addons\script.module.beautifulsoup4\lib\bs4\element.py", line 13, in <module>
                                                from bs4.dammit import EntitySubstitution
                                              File "C:\Users\test\AppData\Roaming\Kodi\addons\script.module.beautifulsoup4\lib\bs4\dammit.py", line 14, in <module>
                                                from htmlentitydefs import codepoint2name
                                            ModuleNotFoundError: No module named 'htmlentitydefs'
                                            -->End of Python script error report<--
2019-10-15 18:37:44.539 T:1320   ERROR: Playlist Player: skipping unplayable item: 0, path [plugin://plugin.video.vrt.nu/play/url/https://www.vrt.be/vrtnu/a-z/vandaag/2019/vandaag-d20191014/]
```
I can workaround this bug by installing [script.module.beautifulsoup4-4.6.3.zip](http://mirrors.kodi.tv/addons/migration/script.module.beautifulsoup4/script.module.beautifulsoup4-4.6.3.zip) manually from zip.

jimfcarroll commented:
@peak3d IIRC you did the reuselanguageinvoker. I'm guessing it was only removed because the branch that got merged was older than your work on reuselanguageinvoker. Can it be re-implemented??

basrieter commented:
I hope so, as it really does speed up a lot of add-ons!

peak3d commented:
Uh, I tend to forget things I have done after a while, but can try to add it again. beside this I remember that we did something in python 2.7 so native libraries can be found on android devices, too (Cryptodome in my case) this is not longer working.

jimfcarroll commented:
Heh ... https://github.com/xbmc/xbmc/pull/13814

peak3d commented:
But how can all theses things get lost just by merging an PR ?

ronie commented:
as far as i can tell it was you who reverted it:
https://github.com/xbmc/xbmc/pull/16116/commits/890f5ac64c6c689ad85f249297cb83909a3fd07f

fuzzard commented:
Well the binmodule patch is a build patch. Would have to have been updated for py3, the reuseinvoker was specifically reverted in a commit in the merge, not sure about the last one ronie mentions

jimfcarroll commented:
Heh, That's what I get for being **volentold** to resolve the conflicts during devcon. :-)

jimfcarroll commented:
Was this something that doesn't work under python3?

basrieter commented:
Given the huge impact of the `reuselanguageinvoker` on performance I would suggest to have another look at it. It would be a shame to loose that benefit.

peak3d commented:
I can't remember reverting it, sry, especially because there was an issue report some days ago of a guy who reports segfaults when the same module is launched twice at the same time. so it should have been in master before the merge of py3. But reading the posts above it seems that revert was done during py3 development, is this correct?

fuzzard commented:
Yes, was part of the feature_python3 branch

jimfcarroll commented:
That revert was part of the Py3 PR that was out there for over a year and had 6-7 commits (one of which was a revert of the reuselanguageinvoker).

jimfcarroll commented:
we can revert the revert and see what happens.

peak3d commented:
can you pls. try @jimfcarroll ? I have an open work place here while updating nx addod to py3

basrieter commented:
If there is a build at some point let me know, I will test it.

Montellese commented:
I can't get this to build on my windows dev machine. The problem is that the `FindPython.cmake` module finds the python executable and python includes installed on my system (`C:\Program Files\Python36\...`) but the library and the `PythonLibs` are found in `project/BuildDependencies/...`. Since they don't match in version it doesn't work.

Python seems to be the only dependency that requires files that are also installed in other places on my machine. I checked and `CMAKE_PREFIX_PATH` is properly set but I'm not sure how to tell `find_program()` and `find_path()` to first check `project/BuildDependencies/...` before checking system paths.

jimfcarroll commented:
I don't know if it's an option for you but @Rechi suggested uninstalling the separately installed python. Obviously this will need to be fixed.

DaveTBlake commented:
Yes @Montellese I hit the same issue on Win dev env. Rechi suggested I uninstall Python that I had separately installed, and that allowed me to compile Kodi in VS2017 but a better long term solution is needed

Montellese commented:
@jimfcarroll @DaveTBlake yeah obviously that works but that's not an acceptable long term solution.

peak3d commented:
@DaveTBlake isn't the key setting PYTHONPATH during kodi startup to something kodi related? empty maybe?
